newscoop/src/Newscoop/NewscoopBundle/EventListener/LocaleListener.php

```diff
@@ -47,62 +55,60 @@ public function onResponse(FilterResponseEvent $event)
+            if ($request->query->has('__set_lang')) {
 
-                    return;
-                }
-            } catch (\Exception $e) {}
-        }
+                try {
+                    $language = $this->em->getRepository('Newscoop\Entity\Language')
+                        ->findByRFC3066bis($request->query->get('__set_lang'), true);
+
+                    if ($language) {
+                        $languageCode = $language->getCode();
+                    }
+                } catch (\Exception $e) {}
+            }
 
-        // if it's not admin
-        if ($pos === false) {
-            // try to get locale from issue
+            // Try to get locale from issue
             $issueMetadata = $request->attributes->get('_newscoop_issue_metadata');
             $issueLanguageCode = $issueMetadata['code_default_language'];
             if ($issueLanguageCode) {
-                $request->setLocale($issueLanguageCode);
-
-                return;
+                $languageCode = $issueLanguageCode;
             }
 
-            // try to get locale from publication but only if locale is not in url
-            $publicationMetadata = $request->attributes->get('_newscoop_publication_metadata');
-            $languageCode = $publicationMetadata['publication']['code_default_language'];
+            // Determine language from URL
             $locale = $this->extractLocaleFromUri($request->getRequestUri());
-            if (!$locale) {
-                $request->setLocale($languageCode);
-
-                return;
+            if ($locale) {
+                $language = $this->em->getRepository('Newscoop\Entity\Language')
+                        ->findOneByCode($locale);
+                if ($language) {
+                    $languageCode = $language->getCode();
+                }
             }
 
-            // get langauge from url and check if it's valid - if it's then set is as current locale
-            $cacheKey = $this->cacheService->getCacheKey(array('resolver', $locale), 'language');
-            if ($this->cacheService->contains($cacheKey)) {
-                $language = $this->cacheService->fetch($cacheKey);
-            } else {
-                $language = $this->em->getRepository('Newscoop\Entity\Language')->findOneBy(array(
-                    'code' => $locale,
-                ));
-
-                $this->cacheService->save($cacheKey, $language);
+            // Last fallback is to use default publication language
+            if (is_null($languageCode)) {
+                $publicationMetadata = $request->attributes->get('_newscoop_publication_metadata');
+                $languageCode = $publicationMetadata['publication']['code_default_language'];
             }
-
-            $request->setLocale(!is_null($language) ? $language->getCode() : $languageCode);
         }
+
+        $request->setLocale($languageCode);
+        $this->session->set('language', $languageCode);
```

Comment:
You are right - saving language to session should be done only for not admin pages. And i'm -1 for `_locale` as it's to similar to request `locale` and we can mix them later. As we save `$languageCode` then in session it can be named - `currentLanguageCode` (or `languageCode`) maybe?